# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [3]:
#!pip install -U scipy

In [4]:
# import numpy and pandas
import numpy as np
import pandas as pd

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [5]:
# Your code here:
data = pd.read_csv("../data/Salaries.csv")

C:\Users\Kate\AppData\Local\Temp\ipykernel_8228\2185703187.py:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("../data/Salaries.csv")


Examine the `salaries` dataset using the `head` function below.

In [6]:
# Your code here:

In [7]:
data.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [8]:
# Your code here:

In [9]:
data.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [10]:
# Your code here

In [11]:
data.shape

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [12]:
# Your code here

In [13]:
data = data.drop(["Notes"], axis=1)
data.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [14]:
# Your code here

In [15]:
data["Status"].value_counts(dropna=False)

NaN    110535
FT      22334
PT      15785
Name: Status, dtype: int64

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [16]:
# Your code here

In [17]:
data = data.dropna(subset = ["Status"])

In [18]:
data.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [19]:
# Your code here

In [20]:
data.dtypes

Id                    int64
EmployeeName         object
JobTitle             object
BasePay              object
OvertimePay          object
OtherPay             object
Benefits             object
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
Agency               object
Status               object
dtype: object

In [21]:
# Your code here

Do any type conversions and reset the index.

In [22]:
# Your code here

In [23]:
data = data.astype({'BasePay':'float','OvertimePay':'float', 'OtherPay':'float', 'Benefits':'float' })
data.reset_index()

index      Id          EmployeeName                     JobTitle  \
0      110531  110532           David Shinn               Deputy Chief 3   
1      110532  110533            Amy P Hart            Asst Med Examiner   
2      110533  110534  William J Coaker Jr.     Chief Investment Officer   
3      110534  110535        Gregory P Suhr              Chief of Police   
4      110535  110536  Joanne M Hayes-White       Chief, Fire Department   
...       ...     ...                   ...                          ...   
38114  148645  148646      Carolyn A Wilson    Human Services Technician   
38115  148647  148648        Joann Anderson  Communications Dispatcher 2   
38116  148648  148649           Leon Walker                    Custodian   
38117  148649  148650         Roy I Tillery                    Custodian   
38118  148653  148654             Joe Lopez   Counselor, Log Cabin Ranch   

         BasePay  OvertimePay   OtherPay  Benefits   TotalPay  \
0      129150.01         0.00  342802.63  38780.04  471952.64   
1      318835.49     10712.95   60563.54  89540.23  390111.98   
2      257340.00         0.00   82313.70  96570.66  339653.70   
3      307450.04         0.00   19266.72  91302.46  326716.76   
4      302068.00         0.00   24165.44  91201.66  326233.44   
...          ...          ...        ...       ...        ...   
38114       0.00         0.00       0.00      0.00       0.00   
38115       0.00         0.00       0.00      0.00       0.00   
38116       0.00         0.00       0.00      0.00       0.00   
38117       0.00         0.00       0.00      0.00       0.00   
38118       0.00         0.00    -618.13      0.00    -618.13   

       TotalPayBenefits  Year         Agency Status  
0             510732.68  2014  San Francisco     PT  
1             479652.21  2014  San Francisco     FT  
2             436224.36  2014  San Francisco     PT  
3             418019.22  2014  San Francisco     FT  
4             417435.10  2014  San Francisco     FT  
...                 ...   ...            ...    ...  
38114              0.00  2014  San Francisco     PT  
38115              0.00  2014  San Francisco     PT  
38116              0.00  2014  San Francisco     PT  
38117              0.00  2014  San Francisco     PT  
38118           -618.13  2014  San Francisco     PT  

[38119 rows x 13 columns]

In [24]:
# Your code here

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [25]:
# Your code here

In [26]:
data["Sum_of_Pays"] = False

for index, row in data.iterrows():
    if row["BasePay"]+row["OvertimePay"]+row["OtherPay"]+row["Benefits"] == row["TotalPayBenefits"]:
        data.loc[index, "Sum_of_Pays"] = True

data.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.00  342802.63  38780.04  471952.64         510732.68  2014   
110532     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.00   82313.70  96570.66  339653.70         436224.36  2014   
110534         0.00   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.00   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  Sum_of_Pays  
110531  San Francisco     PT         True  
110532  San Francisco     FT        False  
110533  San Francisco     PT         True  
110534  San Francisco     FT        False  
110535  San Francisco     FT         True

What is the percetage of employees for which the previous assumption is not True?

In [27]:
# Your code here:

In [28]:
data["Sum_of_Pays"].value_counts(normalize=True).mul(100).round(2)

True     73.71
False    26.29
Name: Sum_of_Pays, dtype: float64

There are different departments in the city. List all departments and the count of employees in each department.

In [29]:
# Your code here:

In [30]:
# There is no column indicationg a department in the dataframe. If there was, I would use smth like this:
# data.groupby(["Department"].agg({'Id':count})).reset_index

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [31]:
# Your code here: (compute the "Hourly_Wage")

In [32]:
data["HourlyWage"] = (data["TotalPayBenefits"]/2000)
data.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.00  342802.63  38780.04  471952.64         510732.68  2014   
110532     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.00   82313.70  96570.66  339653.70         436224.36  2014   
110534         0.00   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.00   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  Sum_of_Pays  HourlyWage  
110531  San Francisco     PT         True  255.366340  
110532  San Francisco     FT        False  239.826105  
110533  San Francisco     PT         True  218.112180  
110534  San Francisco     FT        False  209.009610  
110535  San Francisco     FT         True  208.717550

In [33]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)

In [34]:
mean_hourly_wage_ft = data.loc[data["Status"] == "FT", "HourlyWage"].mean()
mean_hourly_wage_ft

69.264206407271

In [35]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees

In [36]:
filtered_ft = data[data["Status"] == "FT"]
ft_array = filtered_ft["HourlyWage"].values
ft_array

array([239.826105, 209.00961 , 208.71755 , ...,  20.595055,  18.337755,
        15.986405])

In [89]:
t = (np.mean(ft_array) - 75)/(np.std(ft_array, ddof=1)/np.sqrt(len(ft_array)))
print("The value of the statistic is: ",t)

The value of the statistic is:  -35.80631941460526


##### Our statistics is -35.806, and the p-value is so small that it is represented in scientific notations (it is very close to 0)

In [83]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
import scipy.stats as st

In [86]:
lower_critical_value, upper_critical_value = st.t.ppf((0.05/2), df=len(ft_array)-1), st.t.ppf(1-(0.05/2), df=len(ft_array)-1)
print("The lower critical value is {:.3f}".format(lower_critical_value))
print("The upper critical value is {:.3f}".format(upper_critical_value))

The lower critical value is -1.960
The upper critical value is 1.960


##### Our statistics is -35.806 and it is not within the critical values. This means that we are rejecting the H0 - the hourly wage is significantly different from $75/hr.

In [ ]:
# Method 2: Use the p-value method.
# Your code here:

In [92]:
st.t.cdf(t, df=len(ft_array)-1)

2.2589156169197237e-273

##### For H0 to be accepted we need pval to be within the alpha/2 and 1- alpha/2). But in out case it is not True, so we reject H0

In [ ]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [91]:
from scipy.stats import ttest_1samp
stat, pval = ttest_1samp(ft_array, popmean = 75, alternative = "two-sided")
stat, pval

(-35.80631941460526, 4.517831233839447e-273)

Are all the methods in agreement?

##### Here we are getting the same values as in the previous methods. All the methods are in agreement.

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [94]:
police_emp = lambda job: "Police" in job
filtered_plc = data[data["JobTitle"].apply(police_emp)]
plc_array = filtered_plc["TotalPayBenefits"].values
np.mean(plc_array)

148349.04212519727

In [ ]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees

In [95]:
t = (np.mean(plc_array) - 86000)/(np.std(plc_array, ddof=1)/np.sqrt(len(plc_array)))
print("The value of the statistic is: ",t)

The value of the statistic is:  50.252984742109945


In [ ]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

In [ ]:
critical_value = scipy.stats.t.ppf(0.05, df=len(plc_array)-1)
print("The critical value is {:.3f}".format(critical_value))

##### Our statistics is bigger than a critical value, tehrefore we accept H0, salaries this year are significantly higher.

In [ ]:
# Method 2: Use the p-value method.
# Your code here:

In [96]:
scipy.stats.t.cdf(t, df=len(plc_array)-1)

1.0

In [ ]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [97]:
stat, pval = ttest_1samp(plc_array, popmean = 86000, alternative = "less") 
stat, pval

(50.252984742109945, 1.0)

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [98]:
data["JobTitle"].value_counts()

Transit Operator                2479
Special Nurse                   1478
Registered Nurse                1234
Public Svc Aide-Public Works     916
Firefighter                      815
                                ... 
Claims Process Clerk               1
Asst Dir, Log Cabin Rnch           1
Media Production Specialist        1
Chf Prob Ofc, Juv Court            1
Electric Motor Repair Sprv 1       1
Name: JobTitle, Length: 996, dtype: int64

In [ ]:
# Your code here: (Get the department which has most employees)

In [99]:
filtered_to = data[data["JobTitle"] == "Transit Operator"]
to_array = filtered_to["HourlyWage"].values
to_array
np.mean(to_array)

43.359497739007665

In [ ]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees

In [100]:
t = (np.mean(to_array) - 35)/(np.std(to_array, ddof=1)/np.sqrt(len(to_array)))
print("The value of the statistic is: ",t)

The value of the statistic is:  19.064990324906383


In [ ]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

In [101]:
critical_value = scipy.stats.t.ppf(1-0.05, df=len(to_array)-1)
print("The critical value is {:.3f}".format(critical_value))

The critical value is 1.645


##### Out statistics is higher than a critical value, means we can reject the H0, the salaries are not lower than 35 USD/hour.

In [ ]:
# Method 2: Use the p-value method.
# Your code here:

In [103]:
st.t.cdf(t,df=len(to_array)-1)

1.0

In [ ]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [104]:
stat, pval = ttest_1samp(to_array, popmean = 35, alternative = "greater") 
stat, pval

(19.064990324906386, 5.012771694203921e-76)

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [ ]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:

Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [ ]:
# Your code here:

# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [49]:
# Your code here: (Create the function)
def get_department(x):
    if "Police" in x:
        return "Policemen"
    elif "Fire" in x:
        return "Firemen"
    elif "Med" in x or "Nurse" in x:
        return "Medical"
    elif "Attorney" in x:
        return "Lawyer"
    else:
        return "Other"

In [50]:
# Your code here: (Apply the function)

In [51]:
data["employee_group"] = data["JobTitle"].apply(get_department)
data.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.00  342802.63  38780.04  471952.64         510732.68  2014   
110532     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.00   82313.70  96570.66  339653.70         436224.36  2014   
110534         0.00   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.00   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  Sum_of_Pays  HourlyWage employee_group  
110531  San Francisco     PT         True  255.366340          Other  
110532  San Francisco     FT        False  239.826105        Medical  
110533  San Francisco     PT         True  218.112180          Other  
110534  San Francisco     FT        False  209.009610      Policemen  
110535  San Francisco     FT         True  208.717550        Firemen

Determine how many "PT" and "FT" employess have all the employees groups.

In [ ]:
# Your code here: (Store the output dataframe into a new variable)

In [52]:
data_crosstab = pd.crosstab(data['employee_group'], data['Status'])
data_crosstab

Status             FT     PT
employee_group              
Firemen          1333    178
Lawyer            317    102
Medical          1028   2889
Other           18126  12245
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [ ]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:

In [55]:
df = pd.DataFrame()

In [ ]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"

In [61]:
firemen_ft = ((1333+178)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((1333+317+1028+18126+1530)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
firemen_ft

885.2979878800596

In [ ]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"

In [62]:
firemen_pt = ((1333+178)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((178+102+2889+12245+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
firemen_pt

625.7020121199403

In [ ]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"

In [63]:
lawyers_ft = ((317+102)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((1333+317+1028+18126+1530)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
lawyers_ft

245.49295626852748

In [ ]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"

In [65]:
lawyers_pt = ((317+102)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((178+102+2889+12245+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
lawyers_pt

173.5070437314725

In [ ]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"

In [66]:
medical_ft = ((1028+2889)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((1333+317+1028+18126+1530)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
medical_ft

2294.9783047823917

In [ ]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"

In [67]:
medical_pt = ((1028+2889)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((178+102+2889+12245+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
medical_pt

1622.021695217608

In [ ]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"

In [68]:
other_ft = ((18126+12245)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((1333+317+1028+18126+1530)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
other_ft

17794.430966184842

In [ ]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"

In [69]:
other_pt = ((18126+12245)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((178+102+2889+12245+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
other_pt

12576.569033815158

In [ ]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"

In [70]:
policemen_ft = ((1530+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((1333+317+1028+18126+1530)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
policemen_ft

1113.7997848841785

In [ ]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"

In [71]:
policemen_pt = ((1530+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*((178+102+2889+12245+371)/(1333+178+317+102+1028+2889+18126+12245+1530+371))*38119
policemen_pt

787.2002151158215

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [ ]:
# Your code here:

In [79]:
FT_freq = [firemen_ft, lawyers_ft, medical_ft, other_ft, policemen_ft]
PT_freq = [firemen_pt, lawyers_pt, medical_pt, other_pt, policemen_pt]
freq_dict = {'FT': FT_freq, 'PT': PT_freq}
frequencies = pd.DataFrame.from_dict(freq_dict)
frequencies

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [ ]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.05

In [80]:
import scipy.stats as st
st.chi2_contingency(frequencies)

Chi2ContingencyResult(statistic=8.141292346545486e-28, pvalue=1.0, dof=4, expected_freq=array([[  885.29798788,   625.70201212],
       [  245.49295627,   173.50704373],
       [ 2294.97830478,  1622.02169522],
       [17794.43096618, 12576.56903382],
       [ 1113.79978488,   787.20021512]]))

In [81]:
st.chi2.ppf(0.95,df=4)

9.487729036781154

##### Since our statistics (8.1412) is lower than critical value (9.4877), we accept H0 and admit that Status and Department are independent.

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

##### Yes, they are the same!